### ISSUE TITLE
Unexpected handling of binary annotations containing signed long values 

### ISSUE THREAD

**Issue by tristanpenman** [openzipkin/zipkin#1268]
I noticed after upgrading from version 1.7 to 1.8 (and later 1.8.1) of Zipkin, that the recent Codec changes may have introduced a regression in the handling of binary annotations containing negative ints or longs.

I found it easy to reproduce using the TracingBasics demo in [activator-akka-tracing](https://github.com/levkhomich/activator-akka-tracing) which records the result of `Random.nextLong()` as a binary annotation.

After running this demo, attempting to 'Find Traces' using the UI results in an error similar to:

```
Error executing query: Expected -1005656679588439279 to be unsigned
```

I've been able to trace this back to the checkArgument call in Buffer.java:246

Admittedly, I don't understand Zipkin well enough to know whether this was/is now intended behaviour. If it is, I can raise an issue with the authors of the akka-tracing / activator-akka-tracing projects.

**Comment by codefromthecrypt** [comment]
this probably wasn't caught because people don't use Long annotations (as
they are unqueryable and likely to be removed in the future).

That said, happy to fix the regression.. will have a go shortly!

**Comment by codefromthecrypt** [comment]
in general, tracers toString inputs like numbers, which allows them to be
queryable.

here's the model v2 thing which hopefully will remove the niche numeric
types like i16 https://github.com/openzipkin/zipkin/issues/939

**Comment by tristanpenman** [comment]
You might have set a record for fastest issue response there. Thanks for looking into it!

So it sounds like the right thing for me to do here would be to raise issues (or submit PRs) for the projects that I'm using, since string values are clearly the direction forward.

**Comment by codefromthecrypt** [comment]
oh interesting.. I think that there's an issue with B3 in that demo. Ex.
trace ids are being toString'ed as opposed to lower-hex.

you can copy/paste stuff from here until we have extracted a  B3 library

https://github.com/uber/jaeger-client-java/blob/master/jaeger-zipkin/src/main/java/com/uber/jaeger/propagation/b3/B3TextMapCodec.java

**Comment by codefromthecrypt** [comment]
I can't find that line number (with checkArgument), but I can add some
tests to flush out a bug hopefully.

**Comment by tristanpenman** [comment]
It might help if I paste the code block and path.

/zipkin/zipkin/src/main/java/zipkin/internal/Buffer.java:

```
Buffer writeAscii(long v) {
    if (v == 0) return writeByte('0');
    checkArgument(v > 0, "Expected %s to be unsigned", v);
    int width = asciiSizeInBytes(v);
    int pos = this.pos += width; // We write backwards from right to left.
    while (v != 0) {
      int digit = (int) (v % 10);
      buf[--pos] = HEX_DIGITS[digit];
      v /= 10;
    }
    return this;
  }
```

**Comment by codefromthecrypt** [comment]
just found it, but thanks for the assist!!

**Comment by codefromthecrypt** [comment]
I was on a wrong branch :)

**Comment by codefromthecrypt** [comment]
gonna take me an hour (developer hour.. so maybe a couple :) ) to fix
this, as I intentionally special-cased and forgot that this function
was used for binary annotations. stand by!

**Comment by tristanpenman** [comment]
No problem at all. This response has already exceeded all of my expectations! 👍

**Comment by tristanpenman** [comment]
I'm pleased to say that your fix in commit 3b8d56ac1b817729517e32c7a3d311b7ceaa2fb7 has resolved the issue.

Thanks again for your help!

**Comment by codefromthecrypt** [comment]
Thanks for your help identifying this! 1.8.2 will be out shortly
